compiler/rustc_parse/src/parser/path.rs

```diff
@@ -220,6 +238,29 @@ impl<'a> Parser<'a> {
             segments.push(segment);
 
             if self.is_import_coupler() || !self.eat(&token::ModSep) {
+                if style == PathStyle::Expr
+                    && self.may_recover()
+                    && self.token == token::Colon
+                    && self.look_ahead(1, |token| token.is_ident() && !token.is_reserved_ident())
```

Comment:
I was originally responsible for this recovery in my initial PR but I really don't like it much. I _think_ it's fine given the restrictions but it does recovery on the good path which is not good. I would really prefer it if we could avoid this here, although it will probably be fine (probably).